gax/src/test/java/com/google/api/gax/retrying/ContextAwareScheduledRetryingExecutorTest.java

```diff
@@ -0,0 +1,366 @@
+          exception = (CustomException) e.getCause();
+        }
+        assertTrue(attemptResult.isDone());
+        if (!future.cancel(true)) {
+          failedCancelations++;
+        }
+      } while (exception != null && checks < maxRetries);
+
+      assertTrue(future.isDone());
+      assertNotNull(cancellationException);
+      // future.cancel(true) may return false sometimes, which is ok. Also, the every cancellation
+      // of
+      // an already cancelled future should return false (this is what -1 means here)
+      assertEquals(2, checks - (failedCancelations - 1));
+      assertTrue(future.getAttemptSettings().getAttemptCount() > 0);
+      assertFutureCancel(future);
+      localExecutor.shutdownNow();
+    }
+  }
+
+  @Test
+  public void testCancelOuterFutureAfterStart() throws Exception {
+    for (int executionsCount = 0; executionsCount < EXECUTIONS_COUNT; executionsCount++) {
+      ScheduledExecutorService localExecutor = Executors.newSingleThreadScheduledExecutor();
+      FailingCallable callable = new FailingCallable(4, "SUCCESS", tracer);
+      RetrySettings retrySettings =
+          FAST_RETRY_SETTINGS
+              .toBuilder()
+              .setInitialRetryDelay(Duration.ofMillis(1_000L))
+              .setMaxRetryDelay(Duration.ofMillis(1_000L))
+              .setTotalTimeout(Duration.ofMillis(10_0000L))
+              .build();
+      RetryingExecutorWithContext<String> executor =
+          getRetryingExecutor(getAlgorithm(FAILING_RETRY_SETTINGS, 0, null), localExecutor);
+      RetryingFuture<String> future =
+          executor.createFuture(
+              callable,
+              FakeCallContext.createDefault().withTracer(tracer).withRetrySettings(retrySettings));
+      future.setAttemptFuture(executor.submit(future));
+
+      Thread.sleep(30L);
+
+      boolean res = future.cancel(false);
+      assertTrue(res);
+      assertFutureCancel(future);
+      assertTrue(future.getAttemptSettings().getAttemptCount() < 4);
+      localExecutor.shutdownNow();
+    }
+  }
+
+  @Test
+  public void testCancelIsTraced() throws Exception {
+    ScheduledExecutorService localExecutor = Executors.newSingleThreadScheduledExecutor();
+    FailingCallable callable = new FailingCallable(4, "SUCCESS", tracer);
+    RetrySettings retrySettings =
+        FAST_RETRY_SETTINGS
+            .toBuilder()
+            .setInitialRetryDelay(Duration.ofMillis(1_000L))
+            .setMaxRetryDelay(Duration.ofMillis(1_000L))
+            .setTotalTimeout(Duration.ofMillis(10_0000L))
+            .build();
+    RetryingExecutorWithContext<String> executor =
+        getRetryingExecutor(getAlgorithm(FAILING_RETRY_SETTINGS, 0, null), localExecutor);
+    RetryingFuture<String> future =
+        executor.createFuture(
+            callable,
+            FakeCallContext.createDefault().withTracer(tracer).withRetrySettings(retrySettings));
+    future.setAttemptFuture(executor.submit(future));
+
+    Thread.sleep(30L);
```

Comment:
DANGER! This adds 30 milliseconds to every test run. Can we not do that?

Reply:
Good point. I've replaced it with a latch (plus a very short busy wait).